discojs/discojs-core/src/aggregator/base.ts

```diff
@@ -0,0 +1,239 @@
+
+export enum AggregationStep {
+  ADD,
+  UPDATE,
+  AGGREGATE
+}
+
+export abstract class Base<T> {
+  /**
+   * Contains the ids of all active nodes, i.e. members of the aggregation group at
+   * a given round. It is a subset of all the nodes available in the network.
+   */
+  protected _nodes: Set<client.NodeID>
+  /**
+   * Contains the contributions received from active nodes, accessible
+   * by node id. It defines the effective aggregation group, which is
+   * possibly a subset of all active nodes, depending on the aggregation
+   * scheme.
+   */
+  protected contributions: Map<number, Map<client.NodeID, T>>
+  /**
+   * Emits the aggregation event whenever an aggregation step is performed.
+   * Triggers the resolve of the result promise and the preparation for the
+   * next aggregation round.
+   */
+  private readonly eventEmitter: EventEmitter
+
+  protected informant?: AsyncInformant<T>
+  /**
+   * The result promise which, on resolve, will contain the current aggregation result.
+   */
+  protected result: Promise<T>
+  /**
+   * The current aggregation round, used for assessing whether a contribution is recent enough
+   * or not.
+   */
+  protected _round = 0
+
+  protected _communicationRound = 0
+
+  constructor (
+    public readonly task: Task,
+    protected _model?: tf.LayersModel,
+    protected readonly roundCutoff = 0,
+    public readonly communicationRounds = 1
+  ) {
+    this.eventEmitter = new EventEmitter()
+    this.contributions = Map()
+    this._nodes = Set()
+
+    this.result = this.makeResult()
+
+    this.eventEmitter.on('aggregation', () => {
+      this.nextRound()
+    })
+  }
+
+  /**
+   * Adds a node's contribution to the aggregator for a given round.
+   * The contribution will be aggregated during the round's aggregation step.
+   * @param nodeId The node's id
+   * @param contribution The node's contribution
+   * @param round For which round the contribution was made
+   */
+  abstract add (nodeId: client.NodeID, contribution: T, round: number, communicationRound?: number): boolean
+
+  /**
+   * Performs the aggregation step over the received node contributions.
+   * Must store the aggregation's result in the aggregator's result promise.
+   */
+  abstract aggregate (): void
+
+  registerObserver (informant: AsyncInformant<T>): void {
+    this.informant = informant
+  }
+
+  /**
+   * Returns whether the given round is recent enough, dependent on the
+   * aggregator's round cutoff.
+   * @param round The round
+   * @returns True if the round is recent enough, false otherwise
+   */
+  isWithinRoundCutoff (round: number): boolean {
+    return this.round - round <= this.roundCutoff
+  }
+
+  /**
+   * Logs useful messages during the various aggregation steps.
+   * @param step The aggregation step
+   * @param from The node which triggered the logging message
+   */
+  log (step: AggregationStep, from?: client.NodeID): void {
+    switch (step) {
+      case AggregationStep.ADD:
```

Comment:
add might be maybe misinterpreted as addition potentially, but here you mean more like `save` or `register`?

Reply:
it corresponds to the aggregator's `add` method and will have a dedicated docstring in #580 so it'll make more sense then :)